zeus/monitor/carbon.py

```diff
@@ -0,0 +1,448 @@
+            gpu_energy=overall_measurement.gpu_energy,
+            cpu_energy=overall_measurement.cpu_energy,
+            dram_energy=overall_measurement.dram_energy,
+            gpu_carbon_emission=self.finished_keys[key][0],
+            cpu_carbon_emission=self.finished_keys[key][1] or None,
+        )
+
+        del self.finished_keys[key]
+
+        return measurement
+
+
+def _polling_process(
+    command_q: mp.Queue,
+    finished_q: mp.Queue,
+    gpu_indices: list[int],
+    cpu_indices: list[int],
+    carbon_intensity_provider: CarbonIntensityProvider,
+):
+    last_index = 0
+    index = 0
+    zeus_monitor = ZeusMonitor(gpu_indices=gpu_indices, cpu_indices=cpu_indices)
+    gpu_carbon_emissions = defaultdict(
+        lambda: defaultdict(float)
+    )  # {window_key -> {gpu index -> cumulative carbon emission}}
+    cpu_carbon_emissions = defaultdict(
+        lambda: defaultdict(float)
+    )  # {window_key -> {cpu index -> cumulative carbon emission}}
+    energy_measurements = defaultdict(
+        lambda: defaultdict(lambda: defaultdict(float))
+    )  # {window_key -> {hour -> {gpu/cpu index -> energy}}}
+    keys = set()
+
+    # record energy measurement
+    def _update_energy_measurements(key: str, hour_key: str):
+        measurement = zeus_monitor.end_window(key)
+
+        for gpu_index, energy_measurement in measurement.gpu_energy.items():
+            energy_measurements[key][hour_key][f"gpu_{gpu_index}"] = energy_measurement
+
+        if measurement.cpu_energy:
+            for cpu_index, energy_measurement in measurement.cpu_energy.items():
+                energy_measurements[key][hour_key][
+                    f"cpu_{cpu_index}"
+                ] = energy_measurement
+
+    # update cumulative carbon emissions
+    def _update_carbon_emissions(key: str):
+        carbon_intensities = carbon_intensity_provider.get_recent_carbon_intensity()
+
+        for hour_key, carbon_intensity in carbon_intensities.items():
+            for gpu_index in zeus_monitor.gpu_indices:
+                # divide by 3600 to convert joules -> Wh
+                gpu_carbon_emissions[key][gpu_index] += (
+                    energy_measurements[key][hour_key][f"gpu_{gpu_index}"]
+                    / 3600
+                    * carbon_intensity
+                )
+
+            for cpu_index in zeus_monitor.cpu_indices:
+                # divide by 3600 to convert joules -> Wh
+                cpu_carbon_emissions[key][cpu_index] += (
+                    energy_measurements[key][hour_key][f"cpu_{cpu_index}"]
+                    / 3600
+                    * carbon_intensity
+                )
+
+        del energy_measurements[key]
+
+    while True:
+        # calculate current time
+        now = datetime.now(timezone.utc)
+        hour_floor = now.replace(minute=0, second=0, microsecond=0)
+        hour_ceil = hour_floor + timedelta(hours=1)
+        hour_key = hour_floor.strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"
```

Comment:
So..... this matches how it's returned from ElectricityMap's JSON response? If so, is this assuming that as part of the `CarbonIntensityProvider` contract? What if another carbon intensity provider provides data in a different time format?

Reply:
Ah did not consider that. This datetime formatting is based on the formatting of ElectricityMaps and definitely will run into issues with other carbon intensity providers. Maybe we can pass in a formatting function to the _polling_process so that other carbon intensity providers can also be used?

Reply:
I think the `CarbonIntensityProvider.get_recent_carbon_intensity`'s return value should be something like a `dict[datetime, float]`. WDYT?

Reply:
ah yes good idea, I will do that